src/types/widget.ts

```diff
@@ -110,102 +111,167 @@ export type BuiltinWidgetTypes =
   | 'ais.toggleRefinement'
   | 'ais.voiceSearch';
 
-/**
- * Widgets are the building blocks of InstantSearch.js. Any valid widget must
- * have at least a `render` or a `init` function.
- */
-export type Widget<
-  TWidgetOptions extends { renderState: unknown } = { renderState: unknown }
-> = {
-  /**
-   * Identifier for official connectors and widgets
-   */
-  $$type?: BuiltinTypes;
+export type WidgetParams = {
+  widgetParams: NonNullable<unknown>;
+};
 
+export type WidgetDescription = {
+  $$type: string;
+  $$widgetType?: string;
+  renderState?: Record<string, unknown>;
+  indexRenderState?: Record<string, unknown>;
+  indexUiState?: Record<string, unknown>;
+};
+
+type RequiredWidgetLifeCycle<TWidgetDescription extends WidgetDescription> = {
   /**
-   * Identifier for official widgets
+   * Identifier for connectors and widgets.
    */
-  $$widgetType?: BuiltinWidgetTypes;
+  $$type: TWidgetDescription['$$type'];
 
   /**
-   * Called once before the first search
+   * Called once before the first search.
    */
-  init?(options: InitOptions): void;
+  init?: (options: InitOptions) => void;
   /**
-   * Called after each search response has been received
+   * Called after each search response has been received.
    */
-  render?(options: RenderOptions): void;
+  render?: (options: RenderOptions) => void;
   /**
    * Called when this widget is unmounted. Used to remove refinements set by
    * during this widget's initialization and life time.
    */
-  dispose?(options: DisposeOptions): SearchParameters | void;
+  dispose?: (options: DisposeOptions) => SearchParameters | void;
+};
+
+type RequiredWidgetType<TWidgetDescription extends WidgetDescription> = {
+  /**
+   * Identifier for widgets.
+   */
+  $$widgetType: TWidgetDescription['$$widgetType'];
+};
+
+type WidgetType<
+  TWidgetDescription extends WidgetDescription
```

Comment:
why not use WidgetDescription directly here?

Reply:
because the actual value of the description is needed, so it needs to be a generic :)